### DISCUSSION TITLE
Slow(?) React native performance (benchmark repo inside)

### DISCUSSION THREAD

**Discussion by efstathiosntonas** [tamagui/tamagui#1471] [General]
Hi folks, 

I've created a [benchmark repo](https://github.com/efstathiosntonas/react-native-style-libraries-benchmark) for all major style libraries out there for react-native.

I've noticed that Tamagui takes 506ms on Avg. to render `App.js`, that's 267% slower than react-native `StyleSheet`. This number is really high, can you please take a look at the repo and check if everything is setup properly? I have a feeling that I am missing something.

Thank you for this amazing library. 

edit: after using `Stack` instead of `styled(View,` the performance difference increased to 327%

**Comment by pehkay** [comment]
I did noticed the performance drop from version 1.36.x vs the current.

**Reply by efstathiosntonas** [reply]
@pehkay I downgraded to 1.36.6 and it takes 600ms on average, it's even worse than 1.45.11

**Reply by pehkay** [reply]
Wow, that is weird. :). Let me try 1.45.11.

**Comment by atanaskanchev** [comment]
That's quite worrisome, [their own benchmarks](https://tamagui.dev/docs/intro/benchmarks) hosted [here](https://github.com/tamagui/benchmarks) (most probably obsolete by now) paint a different picture...

**Comment by efstathiosntonas** [comment]
@atanaskanchev the configuration is not good, the compiler is not kicking in. Trying to follow the official benchmark repo but I stumble on some errors.

**Comment by natew** [comment]
I'll take a look, likely something is configured wrong or we had a regression.

**Reply by natew** [reply]
In fact I can see you're re-rendering the provider on each render, which is a red flag. You'd want to have the provider above the rest otherwise it's not really a fair comparison.

**Reply by efstathiosntonas** [reply]
Hi @natew, thanks for looking into this. Provider is not re-rendering on each loop, it's wrapping the entire component.

**Reply by claudesortwell** [reply]
@natew I tested with the Provider at the top level and still got the same result. Nothing is change in render except the stack, Test is just a component with the switcher state in it.

<img width="783" alt="Screenshot 2023-07-28 at 5 31 26 pm" src="https://github.com/tamagui/tamagui/assets/41422239/57ac47f3-0f76-4091-ba53-d01f668516ad">

**Reply by natew** [reply]
I’ll have a whole write up about it at some point but there are a few things that regressed. Actually Hermes is super sensitive to some patterns and so we’re just discovering that. Super small changes actually sped things up a bunch. 

But also running these tests in dev mode makes them mostly pointless, and they are so arbitrary. A huge piece of why Tamagui is slower is because we add a few more default styles. So for a super tiny micro benchmark like this is looks big. I found two we can remove and that makes it almost the same perf as the others align with two small Hermes related changes. 

Also note that this test case would actually be ideal for the optimizing compiler. You’d get actually the same speed as native. 

Also the test for Tamagui has more styles than the others for some reason. 

Anyway thanks for profiling things. I’m going to use this as an opportunity to revisit performance in general. I see a few more things I can fine tune. The next release out tomorrow or so should do well.

**Reply by efstathiosntonas** [reply]
@natew thank you for taking the time to investigate it. I've updated the repo and the tamagui styles now align with the other's libraries components and the scores did not improve. Looking forward for the tuned release!

**Comment by efstathiosntonas** [comment]
Hey folks, @natew has improved the performance by a lot and as a sidenote: compiler do not kick in in dev mode so performance is greatly improved in release/production mode.

**Reply by Bowlerr** [reply]
Comparing other libraries and solutions are the improvements equal to the statistics that tamagui shows on their website?

**Comment by rommyarb** [comment]
Did you run the test in dev mode? 😄

**Reply by natew** [reply]
yes lol...

also this type of super micro benchmark is not very insightful. for example tamagui expands styles for correctness, so in this case we were expanding borderWidth => borderLeftWidth, etc. turns out that caused most of the performance difference. but in reality, when you're rendering a huge page of things expansions like this shouldn't make nearly as much of a difference.

i've actually brought this up with RN team, but need to file a proper issue

the dangers of micro benchmarks are real.
